password screen for DB

ayoub-benali opened the pull request:
- [x] during the registration ask the user to provide a password used to protect the DB
- [x] ask the user to enter the password when the account exists already

Commit:
WIP add application context

holding database and protocol store

Commit:
initialize database on creation of protocol store

by issueing schema create statements

Commit:
cleanup

Commit:
fix tests

Commit:
WIP: implement registrationdata store and prekey id sequences

Commit:
fix tests

Commit:
use the app context to create store on the finish screen

Commit:
ensure async ui actions to be completed in tests

this should fix a flaky TestFX based test

Commit:
WIP add a scree to unlock the DB

Commit:
fix da shizzle

Commit:
Merge branch 'master' into a/password-screen

Commit:
Merge branch 'a/password-screen' of github.com:ayoub-benali/signal-desktop-client into a/password-screen

Commit:
fix stackoverflow exception

Commit:
mini refactoring

Commit:
Merge branch 'master' into a/password-screen

hsanjuan commented:
Make passwords optional? They don't add so much over an encrypted drive.

ayoub-benali commented:
What do you think @mfelsche  ?

mfelsche commented:
think about the poor windows users...

mfelsche commented:
But yeah, we totally can also allow not using a password, for the ones who love to live dangerously.
This means that when no password is added on registration, the app starts right away into the main screen without asking for password.

Commit:
fix loading the data when the account is already created

Commit:
fix show the chat after loading the database

ayoub-benali commented on src/main/scala/de/m7w3/signal/ApplicationContext.scala:
```diff
@@ -68,7 +67,7 @@ case class DatabaseContext(config: Config.SignalDesktopConfig) {
     * actual path of the database
     */
-  def dbFilePath: Path = Paths.get(uriDbPath.toString, "mv.db")
+  def dbFilePath: Path = Paths.get(uriDbPath.toString + ".mv.db")
```
@mfelsche for some weird reason, the initial way doesn't find the file. So I had to do this way

ayoub-benali commented on src/main/scala/de/m7w3/signal/ApplicationContext.scala:
```diff
@@ -41,7 +41,7 @@ case class ApplicationContext(config: Config.SignalDesktopConfig) {
 
   def tryLoadExistingStore(password: String): Try[SignalDesktopProtocolStore] = {
-    databaseContext.dbActionRunner(password, onlyIfExists = true)
+    databaseContext.dbActionRunner(password, onlyIfExists = false)
```
@mfelsche had to change it to avoid throwing an DatabaseDoesNotExistException [here](https://github.com/ayoub-benali/signal-desktop-client/blob/master/src/main/scala/de/m7w3/signal/ApplicationContext.scala#L83)

ayoub-benali commented:
will probably make the password optional in an other PR

Commit:
fix loading an existing h2 database

Commit:
press the unlock button when ENTER key is pressed

Commit:
remove logging classpath

Commit:
add a flag to override the database var when unlocking the app

the user could type the wrong password there we need to override the
database var

Commit:
use gridpane for the login screen

show an error icon if loading data fails
add some other refinements

Commit:
improve the registration process

Commit:
remove unused imports

Commit:
add and application context to be used during tests

Commit:
fix tabs

mfelsche commented on src/main/scala/de/m7w3/signal/AccountHelper.scala:
```diff
@@ -13,11 +13,10 @@ import org.whispersystems.libsignal.util.{KeyHelper, Medium}
 import org.whispersystems.signalservice.internal.util.Base64
 
-case class AccountHelper(userId: String){
+case class AccountHelper(userId: String, password: String = Util.getSecret(20)){
```
shouldnt this password be somewhat static and not change between each instantiation? So, chose a password once at registration and keep it? using a default argument value here seems way too implicit.

ayoub-benali replied:
It is static. At the next run of the application (after registration) the password from the DB (the same as the first time) is passed here as an argument instead of the default value. 

I can call `Util.getSecret(20)` outside and pass the password explicitly the first time

mfelsche replied:
that would be nice

mfelsche commented on src/main/scala/de/m7w3/signal/ApplicationContext.scala:
```diff
@@ -30,7 +29,7 @@ case class ApplicationContext(config: Config.SignalDesktopConfig) {
 
   def createNewProtocolStore(password: String): SignalDesktopProtocolStore = {
-    databaseContext.dbActionRunner(password, onlyIfExists = false) match {
+    databaseContext.dbActionRunner(password, onlyIfExists = false, false) match {
```
by name pls

ayoub-benali replied:
will do :)

mfelsche commented on src/main/scala/de/m7w3/signal/DeviceRegistration.scala:
```diff
@@ -2,72 +2,125 @@ package de.m7w3.signal
+      finish.disable = true
+      finish.onAction = (a: ActionEvent) => Platform.runLater{
+        // TODO show a progress bar
+        Main.store = context.createNewProtocolStore(Step2.dbPassword.getText())
```
store should be a property of ApplicationContext or something along those lines, same for account

ayoub-benali replied:
I think the same yes. Should we do it in this PR ?

mfelsche replied:
we do it in the next one

mfelsche commented on src/test/scala/de/m7w3/signal/ApplicationContextTest.scala:
```diff
@@ -0,0 +1,7 @@
+import scala.concurrent.duration._
+import java.io.File
+
+object ApplicationContextTest extends ApplicationContext(SignalDesktopConfig(false, 1.seconds, new File("foo")))
```
would rather create a trait that extends beforeAndAfter... and contains an ApplicationContext pointing to some tempdir that is deleted afterAll

ayoub-benali replied:
yeah this thing is still a WIP

mfelsche replied:
just sayin'

Commit:
remove default param value and make param explicit

mfelsche reviewed (approved):
boom
